src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -236,20 +201,57 @@ public Task WriteEndBatchAsync()
         }
 #endif
 
-        /// <summary>Starts a new changeset; can only be called after WriteStartBatch and if no other active operation or changeset exists.</summary>
+        /// <summary>
+        /// Starts a new changeset without specifying group id.
+        /// This can only be called after WriteStartBatch and if no other active operation or changeset exists.
+        /// </summary>
         public void WriteStartChangeset()
         {
+            this.WriteStartChangeset(Guid.NewGuid().ToString());
+        }
+
+        /// <summary>
+        /// Starts a new atomic group or changeset with the specified group id or changeset GUID corresponding to change set boundary.
+        /// This can only be called after WriteStartBatch and if no other active operation or changeset exists.</summary>
+        /// <param name="groupOrChangesetId">
+        /// The atomic group id (for Json batch) / changeset GUID (for Multipart/Mixed batch) of the batch request.
+        /// Cannot be null.
+        /// </param>
+        /// <exception cref="ODataException">Thrown if the <paramref name="groupOrChangesetId"/> is null.</exception>
+        public void WriteStartChangeset(string groupOrChangesetId)
+        {
+            ExceptionUtils.CheckArgumentNotNull(groupOrChangesetId, "groupOrChangesetId");
+
             this.VerifyCanWriteStartChangeset(true);
-            this.WriteStartChangesetImplementation();
+            this.WriteStartChangesetImplementation(groupOrChangesetId);
+            this.FinishWriteStartChangeset();
         }
 
 #if PORTABLELIB
-        /// <summary>Asynchronously starts a new change set; can only be called after WriteStartBatch and if no other active operation or change set exists.</summary>
+        /// <summary>Asynchronously starts a new change set without specifying group id;
+        /// This can only be called after WriteStartBatch and if no other active operation or change set exists.</summary>
         /// <returns>A task instance that represents the asynchronous write operation.</returns>
         public Task WriteStartChangesetAsync()
         {
+            return WriteStartChangesetAsync(Guid.NewGuid().ToString());
+        }
+
+        /// <summary>
+        /// Asynchronously starts a new change set; can only be called after WriteStartBatch and if no other active operation or change set exists.
+        /// </summary>
+        /// <param name="groupOrChangesetId">
+        /// The atomic group id (for Json batch) / changeset GUID (for Multipart/Mixed batch) of the batch request.
+        /// Cannot be null.
+        /// </param>
+        /// <returns>A task instance that represents the asynchronous write operation.</returns>
+        /// <exception cref="ODataException">Thrown if the <paramref name="groupOrChangesetId"/> is null.</exception>
+        public Task WriteStartChangesetAsync(string groupOrChangesetId)

```

Comment:
>groupOrChangesetId [](start = 52, length = 18)

From an API perspective, which should be format-agnostic, this is the id associated with the changeset.  Can we just call this "changesetId" and remove the format-specific comments in the <param> description?

Reply:
updated. Will push the commit.

---
In reply to: [157302410](https://github.com/OData/odata.net/pull/992#discussion_r157302410) [](ancestors = 157302410)